core/src/main/java/org/infinispan/remoting/transport/AbstractRequest.java

```diff
@@ -0,0 +1,113 @@
+package org.infinispan.remoting.transport;
+
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.remoting.responses.Response;
+
+/**
+ * A remote invocation request.
+ *
+ * <p>Thread-safety: This class and its sub-classes are thread-safe. They use the {@code ResponseCollector}'s monitor
+ * for synchronization, so that collectors usually don't need any explicit synchronization.</p>
+ *
+ * @author Dan Berindei
+ * @since 9.1
+ */
+public abstract class AbstractRequest<T> extends CompletableFuture<T> implements Callable<Void>, Request<T> {
+   protected final ResponseCollector<T> responseCollector;
```

Comment:
I'd keep it here, mostly for symmetry: if I put it in the subclasses it would be private in SingleTargetRequest, and protected in MultiTargetRequest.